econml/score/rscorer.py

```diff
@@ -0,0 +1,219 @@
+                 random_state=None):
+        self.model_y = clone(model_y, safe=False)
+        self.model_t = clone(model_t, safe=False)
+        self.discrete_treatment = discrete_treatment
+        self.n_splits = n_splits
+        self.categories = categories
+        self.random_state = random_state
+        self.mc_iters = mc_iters
+        self.mc_agg = mc_agg
+
+    def fit(self, y, T, X=None, W=None, sample_weight=None, groups=None):
+        """
+
+        Parameters
+        ----------
+        Y: (n × d_y) matrix or vector of length n
+            Outcomes for each sample
+        T: (n × dₜ) matrix or vector of length n
+            Treatments for each sample
+        X: optional (n × dₓ) matrix
+            Features for each sample
+        W: optional (n × d_w) matrix
+            Controls for each sample
+        sample_weight: optional (n,) vector
+            Weights for each row
+        groups: (n,) vector, optional
+            All rows corresponding to the same group will be kept together during splitting.
+            If groups is not None, the n_splits argument passed to this class's initializer
+            must support a 'groups' argument to its split method.
+
+        Returns
+        -------
+        self
+        """
+        if X is None:
+            raise ValueError("X cannot be None for the RScorer!")
+
+        self.lineardml_ = LinearDML(model_y=self.model_y,
+                                    model_t=self.model_t,
+                                    n_splits=self.n_splits,
+                                    discrete_treatment=self.discrete_treatment,
+                                    categories=self.categories,
+                                    random_state=self.random_state,
+                                    mc_iters=self.mc_iters,
+                                    mc_agg=self.mc_agg)
+        self.lineardml_.fit(y, T, X=None, W=np.hstack([v for v in [X, W] if v is not None]),
+                            sample_weight=sample_weight, groups=groups, cache_values=True)
+        self.base_score_ = self.lineardml_.score_
+        self.dx_ = X.shape[1]
+        return self
+
+    def score(self, cate_model):
+        """
+        Parameters
+        ----------
+        cate_model : instance of fitted BaseCateEstimator
+
+        Returns
+        -------
+        score : double
+            An analogue of the R-square loss for the causal setting.
+        """
+        Y_res, T_res = self.lineardml_._cached_values.nuisances
+        X = self.lineardml_._cached_values.W[:, :self.dx_]
+        sample_weight = self.lineardml_._cached_values.sample_weight
+        if Y_res.ndim == 1:
+            Y_res = Y_res.reshape((-1, 1))
+        if T_res.ndim == 1:
+            T_res = T_res.reshape((-1, 1))
+        effects = cate_model.const_marginal_effect(X).reshape((-1, Y_res.shape[1], T_res.shape[1]))
+        Y_res_pred = np.einsum('ijk,ik->ij', effects, T_res).reshape(Y_res.shape)

```

Comment:
By looking at your notebook example, so even for cate estimator doesn't follow the residual to residual final stage (e.g. metalearners) but has const_marginal_effect, we could still assume this theta(x)*res_t formula in order to compute r score?

Reply:
yeap. That's exactly right. If something claims to be a cate then it should achieve good residual on residual loss even if it's not trained based on that. Rscore is a univoersal scoring algorithm